### DISCUSSION TITLE
Render a 2D Mesh to a texture/image

### DISCUSSION THREAD

**Discussion by bilowik** [bevyengine/bevy#9036] [Q&A] [answered]
This is probably a far more complex question than it seems and may not be feasible. I've been digging through the source trying to find how a Mesh (probably generically as a RenderAsset::PreparedAsset) gets sent to the underlying wgpu device for rendering but have been struggling. 

I've seen the render_to_texture example, which is great for many things, but it involves creating a camera to look at a specific position and rendering what is at that position. To get a clean render of just the one Mesh, would need to move it far away, ensure it is not overlapping with anything else after that move, wait a frame for it to render to the texture. I am wondering if in the case of just a single Mesh, is there a better approach? 

Thanks for taking the time to read this!

**Comment by nicopap** [comment, marked as answer]
I don't understand your question. I think in all this rambling, you forgot to ask precisely what you were looking for :)

Anyway

### 1 - How to render something with one camera but not another

Use the [`RenderLayers`](https://docs.rs/bevy/latest/bevy/render/view/struct.RenderLayers.html) component. You can set your "render-to-texture" camera and what it is supposed to render to a singular `RenderLayer`, separated from all the other entities layers. This way, you'll have a separate view of the object in question

### 2 - How does bevy rendering works

Vast question! I'm going to do a very high level bird's eye view here, otherwise I'll have to write an entire book.

1. **Extract mesh into the render world**: Meshes are an `Asset`, and as you astutely already noted, they are passed to the render world through their `RenderAsset` ~~dark world~~ render world counterpart. What is accessed by the render world is not a `Mesh`, but a `GpuMesh`, the transformation is done at the extraction stage.
2. **Tell wgpu to render the mesh**: Yeah that's it. The bit of code you are looking for is in `bevy_pbr/src/render/mesh.rs`. This was my first contribution to bevy, 2 years ago now!

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_pbr/src/render/mesh.rs#L1251-L1269

3. **Create a render command with `DrawMesh`**: We have the code, but it needs to be ran, where is it ran? Well, first we make `DrawMesh` part of a larger render command:

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_pbr/src/material.rs#L342-L348

4. **Add the command to a render phase**: Bevy's rendering uses [`RenderGraph`]. A graph is made of [`Node`]s, nodes are basically "render passes" that implement [`ViewNode`], which seems to mostly be worried about calling [`RenderPhase<I>`] for a specific `I` that implements [`PhaseItem`]. What's the relationship with `DrawMaterial`? Well, `DrawMaterial` is a `RenderCommand`, which is associated with a specific `PhaseItem`. Here we associate `DrawMaterial` with three different `PhaseItem`s: `Transparent3d`, `Opaque3d` and `AlphaMask3d`. More on this later.

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_pbr/src/material.rs#L194-L199 

5. **Create a node to access the render phase**: To run any rendering code, we need to add it as node to the `RenderGraph`. bevy defines a set of default nodes in `bevy_core_pipeline`, one of which is `MainOpaquePass3dNode`.

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_core_pipeline/src/core_3d/main_opaque_pass_3d_node.rs#L25

6. **In `ViewNode`: create a render pass**: There is a lot of shared boilerplate between different `Node`s. So bevy defines a wrapper type [`ViewNodeRunner`] that does the shared boilerplate. The unique work is implemented in the [`ViewNode`] trait implementation. Generally, nodes create render passes and execute them. Render passes are a webGPU concept, I'm not quite familiar with it.

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_core_pipeline/src/core_3d/main_opaque_pass_3d_node.rs#L68-L102

7. **In `ViewNode`: call `PhaseItem::render` on [`RenderPhase`]**: `RenderPhase` is a component. They are added to the render-world equivalent of cameras. Each view may have several `RenderPhase`s of different types. `RenderPhase` is a collection of `PhaseItem`s, according to the docs, each `PhaseItem` is an "renderable" entity that is visible from the camera, one needs to add them manually in `queue` systems.

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_core_pipeline/src/core_3d/main_opaque_pass_3d_node.rs#L110-L111

8. **In `PhaseItem::render`, call the `draw` method on all the render phase items**: This is complicated  I think I'm lost here. Somehow, the `DrawMesh` ended up in a resource of type `DrawFunctions<I>`. `DrawFunction` stores a `Vec<Box<dyn Draw>>`. `DrawMesh` does not itself implement `Draw`, but it implements `RenderCommand`. The documentation on `RenderCommand` hints that somehow it was associated with a `RenderCommandState` and added to this `DrawFunctions` resource. But I can't identify the code that does that.

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_render/src/render_phase/mod.rs#L77-L91

I find myself in a maze of little twisty passages, and I'm utterly lost. I can't go forward on this. But I think this should close the loop, as `DrawFunctions::draw` calls the `RenderCommands::render` method, which is the function we read at point (2) of this exposé.

https://github.com/bevyengine/bevy/blob/9478432cb9d7ed0db9e5fafc8be148c0ddae35bc/crates/bevy_render/src/render_phase/draw.rs#L260-L272

[`RenderGraph`]: https://docs.rs/bevy/latest/bevy/render/render_graph/struct.RenderGraph.html
[`Node`]: https://docs.rs/bevy/latest/bevy/render/render_graph/trait.Node.html
[`ViewNode`]: https://docs.rs/bevy/latest/bevy/render/render_graph/trait.ViewNode.html
[`PhaseItem`]: https://docs.rs/bevy/latest/bevy/render/render_phase/trait.PhaseItem.html
[`RenderPhase<I>`]: https://docs.rs/bevy/latest/bevy/render/render_phase/struct.RenderPhase.html
[`RenderPhase`]: https://docs.rs/bevy/latest/bevy/render/render_phase/struct.RenderPhase.html
[`ViewNodeRunner`]: https://docs.rs/bevy/latest/bevy/render/render_graph/struct.ViewNodeRunner.html

**Reply by bilowik** [reply]
Apologies for the rambling. The main question I was asking was how the GPUMesh was processed and passed to the underlying wgpu device, the code I was looking was in that first code block, hadn't looked in bevy_pbr yet so that explains why I hadn't found it. The additional context will also be super helpful thanks!

**Reply by re0312** [reply]
thanks ,really useful to me
